### ISSUE TITLE
Custom crash handler

### ISSUE THREAD

**Issue by nickelser** [celluloid/celluloid-pool#8]
Right now, there's a default `trap_exit` for the pool's actors: https://github.com/celluloid/celluloid-pool/blob/master/lib/celluloid/supervision/container/pool.rb#L163.

It would be awesome if there was an option for a custom `trap_exit` handler to deal with the `(actor, reason)` that right now is being discarded (something like `exit_handler` as an `initialize` argument to the pool alongside `size`?).

I could be totally wrong and there's another way to handle this, but `finalize` on the actual actor does not seem to get the `reason.`

**Comment by digitalextremist** [comment]
You can use `trap_exit :handler` where `:handler` is the name of a method, i.e. `def handler(actor, reason); end` ... that does get both the actor who died, and the reason.

**Comment by digitalextremist** [comment]
You can read about this here:

https://github.com/celluloid/celluloid/wiki/Linking

**Comment by nickelser** [comment]
@digitalextremist I don't think that helps here: this is specifically to get the exits for the pooled workers. I am using the pool to avoid doing the manual work of linking to the workers and managing which ones are living/dead.

I looked at how Sidekiq handled it, and Sidekiq manually manages the worker pool explicitly so it can catch the exits of the linked wokers: https://github.com/mperham/sidekiq/blob/master/lib/sidekiq/manager.rb#L17 which basically re-implements the pool behavior (https://github.com/mperham/sidekiq/blob/master/lib/sidekiq/manager.rb#L102).

I tried adding a `trap_exit` to the pooled class, but it didn't seem to do anything (was not called when it died during normal operation). I'm not sure where else I can add the `trap_exit` handler.

If I monkeypatch the Pool class, I get the behavior I want:

```ruby
module Celluloid
  module Supervision
    class Container
      # Manages a fixed-size pool of actors
      # Delegates work (i.e. methods) and supervises actors
      # Don't use this class directly. Instead use MyKlass.pool
      class Pool
        # Spawn a new worker for every crashed one
        def __crash_handler__(actor, reason)
          @mutex.synchronize {
            @busy.delete actor
            @idle.delete actor
            @actors.delete actor
          }

          STDERR.puts "CRASH_HANDLER: #{reason}"
          # would call the exit_handler at the pool level here

          return unless reason

          @idle << __spawn_actor__
          signal :respawn_complete
        end
      end
    end
  end
end
```

**Comment by nickelser** [comment]
(I'm using this now to handle my use case: https://gist.github.com/nickelser/9f2af7cdfcbc62737982)

**Comment by digitalextremist** [comment]
Oh yes, I see your problem. `trap_exit` is used by `Pool` itself to maintain actors, so it does need to be patched to allow for a *secondary* handler as you're demonstrating in your gist.

Please do post a PR with an implementation as you've shown or similar. You do seem to be using the kind of injections being used in `Pool` and `Supervision` -- so nice adaptation.
